1_fetch/src/fetch_nwis.R

```diff
@@ -1,25 +1,82 @@
-#' Fetch all NWIS sites that have groundwater level data during the appropriate time period
-fetch_gw_sites <- function(start_date, end_date, param_cd){
+#' Fetch NWIS sites within the bounding box that have groundwater level data during the appropriate time period
+fetch_gw_sites <- function(target_name, start_date, end_date, bbox, param_cd){
   
-  hucs <- zeroPad(1:21, 2) # all hucs
+  # From NWIS services documentation: The product of the range of latitude range 
+  #   and longitude multiplied cannot exceed 25 degrees. So, we need to cycle through  
+  #   large bounding boxes by dividing into query cells. Using 5 x 5 cells
+  
+  query_bbox_cells <- prep_latlon_cells(bbox[c(1,3)], bbox[c(2,4)], deg_per_slice = 5)
+  n_cells <- nrow(query_bbox_cells)
   
   sites <- c()
-  for(huc in hucs){
-    message(sprintf("Trying HUC %s ...", huc))
+  for(cell in seq_len(n_cells)) {
+    # For CONUS, we expect some to fail with no data since we put a square grid on
+    # top of the whole country, which means some cells are only over the oceans.
+    message(sprintf("Trying cell %s of %s ...", cell, n_cells))
     sites <- tryCatch(
-      whatNWISdata(huc = huc, 
-                   service = "dv", 
-                   startDate = start_date,
-                   endDate = end_date,
-                   parameterCd = param_cd,
-                   statCd = "00003") %>%
+      whatNWISdata(bBox = extract_cell_bbox(query_bbox_cells, cell), 
+                    service = "dv", 
+                    startDate = start_date,
+                    endDate = end_date,
+                    parameterCd = param_cd,
+                    statCd = "00003") %>%
         pull(site_no) %>%
         unique(), 
       error = function(e) return()
     ) %>% c(sites)
   }
   
   return(sites)
+  
+}
+
+#' Function to divide the requested bounding box into NWIS query-able cells
+prep_latlon_cells <- function(long_vals, lat_vals, deg_per_slice) {
```

Comment:
🥇

Reply:
this took more time to do than I care to admit 😀